### PR TITLE
Limit `isdir` calls in `walkpath`

### PR DESCRIPTION
**Pull request by rofinn**
Currently, if `s3_list_objects` return intermediate "directory" objects then we'll need to call `isdir` on the result. Depending on whether these objects include the trailing "/" this `isdir` call may query AWS. These changes attempt to eliminate that requirement by inspect the next element to determine if an object is just a prefix/directory.

### CODE REVIEW

**Comment by rofinn** [comment]
bors try

File: src/AWSS3.jl [hunk 1]
```diff
@@ -427,7 +427,7 @@ Returns an iterator of `Dict`s with keys `Key`, `LastModified`, `ETag`, `Size`,
 `Owner`, `StorageClass`.
 """
 function s3_list_objects(aws::AbstractAWSConfig, bucket, path_prefix=""; delimiter="/", max_items=nothing, kwargs...)
-    return Channel() do chnl
+    return Channel(ctype=LittleDict, csize=128) do chnl
```
Comment (rofinn) [inline src/AWSS3.jl:430]: Using a sized `Channel` does seem to improve performance slightly. 128 seemed small enough to be fine?

File: src/s3path.jl [hunk 2]
```diff
@@ -86,15 +86,33 @@ function S3Path(
 end
 
 # To avoid a breaking change.
-function S3Path(str::AbstractString; config::AbstractAWSConfig=global_aws_config(),
-                version::AbstractS3Version=nothing)
+function S3Path(
```
Comment (rofinn) [inline src/s3path.jl:89]: Cleaned up this function a little as being able to write `S3Path(str; isdirectory=true)` was handy for debugging the performance issues this PR was trying to solve. If folks want I could make this a separate PR?

File: src/s3path.jl [hunk 3]
```diff
@@ -210,45 +228,73 @@ end
 function FilePathsBase.walkpath(fp::S3Path; kwargs...)
     # Select objects with that prefix
     objects = s3_list_objects(fp.config, fp.bucket, fp.key; delimiter="")
+    root = joinpath(fp, "/")
 
     # Construct a new Channel using a recursive internal `_walkpath!` function
-    return Channel(ctype=typeof(fp)) do chnl
-        _walkpath!(fp, fp, Iterators.Stateful(objects), chnl; kwargs...)
+    return Channel(ctype=typeof(fp), csize=128) do chnl
```
Comment (rofinn) [inline src/s3path.jl:234]: Again, sizing the channel had a modest improvement when scanning thousands of S3 objects.

```
julia> @benchmark collect(walkpath($fp)) seconds=60
BenchmarkTools.Trial: 13 samples with 1 evaluation.
 Range (min … max):  4.573 s …    5.472 s  ┊ GC (min … max): 0.22% … 0.19%
 Time  (median):     4.817 s               ┊ GC (median):    0.16%
 Time  (mean ± σ):   4.857 s ± 240.414 ms  ┊ GC (mean ± σ):  0.11% ± 0.10%

  ▁   ▁ ▁▁▁▁     ▁    █    ▁    █                          ▁
  █▁▁▁█▁████▁▁▁▁▁█▁▁▁▁█▁▁▁▁█▁▁▁▁█▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█ ▁
  4.57 s         Histogram: frequency by time         5.47 s <

 Memory estimate: 25.51 MiB, allocs estimate: 469434.
 ```

size=128:
```
julia> @benchmark collect(walkpath($fp)) seconds=60
BenchmarkTools.Trial: 14 samples with 1 evaluation.
 Range (min … max):  4.197 s …    4.756 s  ┊ GC (min … max): 0.22% … 0.17%
 Time  (median):     4.429 s               ┊ GC (median):    0.19%
 Time  (mean ± σ):   4.461 s ± 171.700 ms  ┊ GC (mean ± σ):  0.11% ± 0.10%

  ▁     ▁     █   ▁  ▁▁      ▁    ▁▁             █▁        ▁
  █▁▁▁▁▁█▁▁▁▁▁█▁▁▁█▁▁██▁▁▁▁▁▁█▁▁▁▁██▁▁▁▁▁▁▁▁▁▁▁▁▁██▁▁▁▁▁▁▁▁█ ▁
  4.2 s          Histogram: frequency by time         4.76 s <

 Memory estimate: 25.37 MiB, allocs estimate: 465806.
```

File: src/s3path.jl [hunk 4]
```diff
@@ -210,45 +228,73 @@ end
...
+        _walkpath!(root, root, Iterators.Stateful(objects), chnl; kwargs...)
     end
 end
 
 function _walkpath!(root::S3Path, prefix::S3Path, objects, chnl; topdown=true, onerror=throw, kwargs...)
+    @assert root.isdirectory
```
Comment (rofinn) [inline src/s3path.jl:240]: If folks want I could make these argument errors, but we should never hit them during normal exection. It's mostly an extra internal check to make sure any changes to this function don't break that assumption.

File: src/s3path.jl [hunk 5]
```diff
@@ -210,45 +228,73 @@ end
...
             # Extract the non-root part of the key
-            k = chop(next["Key"], head=length(root.key), tail=0)
+            k = chop(obj["Key"], head=length(root.key), tail=0)
 
-            # Determine the next appropriate child path
-            # 1. Next is a direct descendant of the current prefix (ie: we have a prefix object)
-            # 2. Next is a distant descendant of the current prefix (ie: we don't have prefix objects)
             fp = joinpath(root, k)
             _parents = parents(fp)
-            child, recurse = if last(_parents) == prefix || fp.segments == prefix.segments
```
Comment (rofinn) [inline src/s3path.jl:238]: Previously, we'd always call `isdir` and query AWS when this condition was hit.

File: src/s3path.jl [hunk 6]
```diff
@@ -210,45 +228,73 @@ end
...
+
+            # If the filepath matches our prefix then pop it off and continue
+            # Cause we would have already processed it before recursing
+            child = if prefix.segments == fp.segments
```
Comment (rofinn) [inline src/s3path.jl:259]: We should be able to skip paths that match our `prefix` as we should have added them in our parent call.

File: src/s3path.jl [hunk 7]
```diff
@@ -210,45 +228,73 @@ end
...
-            else
+                continue
+            # If the filpath is a direct descendant of our prefix then check if it
+            # is a directory too
+            elseif last(_parents) == prefix
+                popfirst!(objects)
+                # If our current path is a prefix for the next path then we can assume that
+                # the current path should be a directory without needing to call `isdir`
+                next = Base.peek(objects)
+                is_dir = (next !== nothing && startswith(next["Key"], fp.key)) || isdir(fp)
```
Comment (rofinn) [inline src/s3path.jl:269]: We only call `isdir` when our current path isn't a prefix for the next. This optimization assumes `s3_list_objects` results will continue to be returned in topdown order.

File: src/s3path.jl [hunk 8]
```diff
@@ -210,45 +228,73 @@ end
...
+                # Reconstruct our next object and explicitly specify whether it is a
+                # directory.
+                S3Path(
```
Comment (rofinn) [inline src/s3path.jl:272]: It'd be nice if there was a way to do a copy and set in Julia to just change the `isdirectory` field.

File: test/s3path.jl [hunk 9]
```diff
@@ -23,6 +23,7 @@ function test_s3_constructors(ps::PathSet)
     @test S3Path(bucket_name, "pathset-root/bar/qux/"; isdirectory=true) == ps.qux
     @test S3Path(bucket_name, p"pathset-root/bar/qux"; isdirectory=true) == ps.qux
     @test S3Path(bucket_name, p"/pathset-root/bar/qux"; isdirectory=true) == ps.qux
+    @test S3Path("s3://$bucket_name/pathset-root/bar/qux"; isdirectory=true) == ps.qux
```
Comment (rofinn) [inline test/s3path.jl:26]: This should address #105

**Review by rofinn — COMMENTED** [review]
I can always break these up into smaller PRs, but they seemed like sufficiently small changes that getting one review would be easier?

**Review by omus — COMMENTED** [review]
I haven't had time to do a thorough review yet. It does seem reasonable though

File: src/AWSS3.jl (hunk 1 above)
Comment (oxinabox) [reply src/AWSS3.jl:430]: sounds fine to me

File: src/s3path.jl (hunk 4 above)
Comment (oxinabox) [reply src/s3path.jl:240]: This seems fine to me

File: src/s3path.jl (hunk 8 above)
Comment (oxinabox) [reply src/s3path.jl:272]: Setfield.jl does this, no?

**Review by oxinabox — APPROVED** [review]

File: src/s3path.jl (hunk 8 above)
Comment (rofinn) [reply src/s3path.jl:272]: Cool, I didn't know about that package. I've made an issue to use that, as I think we do this in other code paths as well.

https://github.com/JuliaCloud/AWSS3.jl/issues/207
